servicetalk-concurrent-api/src/main/java/io/servicetalk/concurrent/api/PublishAndSubscribeOnCompletables.java

```diff
@@ -15,96 +15,107 @@
-            //
-            // This operator acts as a boundary that changes the Executor from original to the rest of the execution
-            // chain. If there is already an Executor defined for original, it will be used to offload signals until
-            // they hit this operator.
-            original.subscribeWithSharedContext(
-                    signalOffloader.offloadSubscriber(
-                            contextProvider.wrapCompletableSubscriber(subscriber, contextMap)), contextProvider);
+            // re-wrap the subscriber so that async context is restored during offloading.
+            Subscriber wrapped = contextProvider.wrapCompletableSubscriber(subscriber, contextMap);
+
+            super.handleSubscribe(wrapped, contextMap, contextProvider);
         }
     }
 
-    private static final class SubscribeOn extends OffloadingCompletable {
+    /**
+     * Completable that invokes on the provided executor the following methods:
+     *
+     * <ul>
+     *     <li>All {@link Cancellable} methods.</li>
+     *     <li>The {@link #handleSubscribe(CompletableSource.Subscriber)} method.</li>
+     * </ul>
+     */
+    private static final class SubscribeOn extends TaskBasedAsyncCompletableOperator {
 
         SubscribeOn(final Completable original, final Executor executor) {
-            super(original, mergeAndOffloadSubscribe(original.executor(), executor));
+            super(original, executor);
+        }
+
+        @Override
+        public Subscriber apply(Subscriber subscriber) {
+            return new CompletableSubscriberOffloadedCancellable(subscriber, executor());
         }
 
         @Override
-        void handleSubscribe(final Subscriber subscriber, final SignalOffloader signalOffloader,
+        void handleSubscribe(final Subscriber subscriber,
                              final AsyncContextMap contextMap, final AsyncContextProvider contextProvider) {
-            // This operator is to make sure that we use the executor to subscribe to the Completable that is returned
-            // by this operator.
-            //
-            // Subscription and handleSubscribe are offloaded at subscribe so we do not need to do anything specific
-            // here.
-            //
-            // This operator acts as a boundary that changes the Executor from original to the rest of the execution
-            // chain. If there is already an Executor defined for original, it will be used to offload signals until
-            // they hit this operator.
-            original.subscribeWithSharedContext(subscriber, contextProvider);
+            try {
+                // re-wrap the subscriber so that async context is restored during offloading.
+                Subscriber wrapped = contextProvider.wrapCompletableSubscriberAndCancellable(subscriber, contextMap);
+
+                // offload the remainder of subscribe()
+                LOGGER.trace("Offloading Completable subscribe() on {}", executor);
```

Comment:
do we need this still or can we remove?

Reply:
It appears to be partially redundant, the `wrapCompletableSubscriber()` is also done in the `handleSubscribe()` method.